include/ginkgo/core/stop/residual_norm.hpp

```diff
@@ -110,7 +316,11 @@ class ResidualNorm
  * @ingroup stop
  */
 template <typename ValueType = default_precision>
-class ResidualNormReduction : public ResidualNorm<ValueType> {
```

Comment:
Why are you publicly inheriting from `ResidualNorm`? You are exposing `ResidualNorm::Factory` and `ResidualNorm::build` to the public interface, which should not happen.
I am currently doing some tests with it and will come back to you if this is indeed a big issue or not (and how to fix it).